packages/sources/coinmetrics/src/endpoint/burned.ts

```diff
@@ -1,63 +1,164 @@
+}
+
+const reDate = new RegExp(/^\d{4}-\d{2}-\d{2}$/)
+
+export const calculateBurnedETH = (jobRunID: string, assetMetricsList: AssetMetrics[]): Decimal => {
+  let burnedETH = new Decimal(0)
+  assetMetricsList.forEach((assetMetrics: AssetMetrics) => {
+    let feeTotNTV
+    let revNtv
+    let issTotNtv
+    try {
+      feeTotNTV = new Decimal(assetMetrics.FeeTotNtv)
+      revNtv = new Decimal(assetMetrics.RevNtv)
+      issTotNtv = new Decimal(assetMetrics.IssTotNtv)
+    } catch (error) {
+      throw new AdapterError({
+        jobRunID,
+        message: `Unprocessable asset metrics: ${JSON.stringify(assetMetrics)}, due to: ${error}.`,
+        statusCode: 400,
+      })
+    }
+    if (feeTotNTV.isNaN() || revNtv.isNaN() || issTotNtv.isNaN()) {
+      throw new AdapterError({
+        jobRunID,
+        message: `Unprocessable asset metrics: ${JSON.stringify(
+          assetMetrics,
+        )}, contains metrics that are not a numbers.`,
+        statusCode: 400,
+      })
+    }
+    burnedETH = burnedETH.plus(feeTotNTV.minus(revNtv.minus(issTotNtv)))
+  })
+  return burnedETH
 }
 
 export const execute: ExecuteWithConfig<Config> = async (request, _, config) => {
   const validator = new Validator(request, inputParameters)
   if (validator.error) throw validator.error
 
+  const today = new Date().toISOString().split('T')[0]
+
   const jobRunID = validator.validated.id
-  const asset = validator.overrideSymbol(AdapterName, validator.validated.data.asset)
-  const frequency = validator.validated.data.frequency || DEFAULT_FREQUENCY
-  const pageSize = validator.validated.data.pageSize || DEFAULT_PAGE_SIZE
-  const url = 'timeseries/asset-metrics'
-  const metrics = 'FeeTotNtv,RevNtv,IssTotNtv'
+  const startDate = validator.validated.data.startDate || DEFAULT_START_DATE
+  const endDate = validator.validated.data.endDate || today
+
+  if (startDate && !reDate.test(startDate)) {
+    throw new AdapterError({
+      jobRunID,
+      message: `Invalid 'startDate': ${startDate}. Expected formats is: YYYY-MM-DD.`,
+      statusCode: 400,
+    })
+  }
+
+  if (Date.parse(startDate) > Date.parse(today)) {
+    throw new AdapterError({
+      jobRunID,
+      message: `Invalid 'startDate': ${startDate}. Is greater than today: ${today}.`,
+      statusCode: 400,
+    })
+  }
+
+  if (!reDate.test(endDate)) {
+    throw new AdapterError({
+      jobRunID,
+      message: `Invalid 'endDate': ${endDate}. Expected formats is: YYYY-MM-DD.`,
+      statusCode: 400,
+    })
+  }
+
+  if (Date.parse(endDate) > Date.parse(today)) {
+    throw new AdapterError({
+      jobRunID,
+      message: `Invalid 'endDate': ${endDate}. Is greater than today: ${today}.`,
+      statusCode: 400,
+    })
+  }
+
+  if (Date.parse(startDate) > Date.parse(endDate)) {
+    throw new AdapterError({
+      jobRunID,
+      message: `Invalid date range: 'startDate': ${startDate} cannot be greater than 'endDate': ${endDate}.`,
+      statusCode: 400,
+    })
```

Comment:
How many of these validations are necessary in the adapter vs get caught by the data provider?

Reply:
All of them are caught by the data provider, however:

1. Any meaningful error message is hidden by how the `Requester` [processes failed requests](https://github.com/smartcontractkit/external-adapters-js/blob/develop/packages/core/bootstrap/src/lib/external-adapter/requester.ts#L52).

Below an example of the error response for a given datetime with bad format:
```typescript
{
    "error": {
        "type": "bad_parameter",
        "message": "Bad parameter 'start_time'. Incorrect time format '2021-1013'. Supported formats are 'yyyy-MM-dd', 'yyyyMMdd', 'yyyy-MM-ddTHH:mm:ss', 'yyyy-MM-ddTHHmmss', 'yyyy-MM-ddTHH:mm:ss.SSS', 'yyyy-MM-ddTHHmmss.SSS', 'yyyy-MM-ddTHH:mm:ss.SSSSSS', 'yyyy-MM-ddTHHmmss.SSSSSS', 'yyyy-MM-ddTHH:mm:ss.SSSSSSSSS', 'yyyy-MM-ddTHHmmss.SSSSSSSSS'."
    }
}
```

This error data is in `error.response.data`. As far as I'm aware the only way to improve the vague `Could not reach endpoint: bad request 400` would be something like this in `Requester.request()`:

```typescript
      let response: AxiosResponse<T>
      try {
        response = await axios(config)
      } catch (error) {
        let message: string
        if (error.response) {
          message =
            `Request failed with status code ${error.response.status}. ` +
            `Response data: ${JSON.stringify(error.response.data)}`
        } else {
          message = JSON.stringify(error.message)
        }
        // Request error
        if (n === 1) {
          logger.error(`Could not reach endpoint: ${message}`)
          throw new AdapterError({ message, cause: error })
        }
        return await _delayRetry(`Caught error. Retrying: ${message}`)
      }
```

2. API `start_time` and `end_time` also accept datetime formats (see error message above). However, `FeeTotNtv, RevNtv, IssTotNtv` are requested by `1d` frequency (set at midnight). I think any valid format other than 'yyyy-MM-dd' and 'yyyyMMdd' is misleading, as the API won't complain and the requester may think the time part is being taken into account.

All in all I just wanted to provide a better understanding for both the user and node operator.